packages/react-query/src/__tests__/useMutation.test.tsx

```diff
@@ -1154,4 +1154,60 @@ describe('useMutation', () => {
 
     expect(onError).toHaveBeenCalledWith(mutateFnError, 'todo', undefined)
   })
+
+  it('should not call mutate callbacks for mutations started after unmount', async () => {
+    const onSuccessMutate = jest.fn()
+    const onSuccessUseMutation = jest.fn()
+    const onSettledMutate = jest.fn()
+    const onSettledUseMutation = jest.fn()
+
+    function Page() {
+      const [show, setShow] = React.useState(true)
+      return (
+        <div>
+          <button onClick={() => setShow(false)}>hide</button>
+          {show && <Component />}
+        </div>
+      )
+    }
+
+    function Component() {
+      const mutation = useMutation({
+        mutationFn: async (text: string) => {
+          await sleep(10)
```

Comment:
Do we need to sleep here? I think this might be redundant now we're no longer using an explicit sleep (nice change by the way) to wait for the unmount and mutation to happen.

```suggestion
```

Reply:
not necessarily, but asynchronous mutation / query functions are preferred in most test because otherwise, render cycles might be batched together and you'd expect to have a loading state where you'll have none because of that.

Reply:
I see, that makes sense. Thanks 🙂